.github/workflows/ci.yml

```diff
@@ -18,72 +18,8 @@ permissions:
-    if: "! contains(toJSON(github.event.commits.*.message), '[skip ci]')"
-
-    # Definition of the build matrix
-    strategy:
-      matrix:
-        java: [11, 17]
-        entry:
-          - { mock-maker: 'mock-maker-default', member-accessor: 'member-accessor-default' }
-          - { mock-maker: 'mock-maker-inline', member-accessor: 'member-accessor-module' }
-          - { mock-maker: 'mock-maker-subclass', member-accessor: 'member-accessor-module' }
-          - { mock-maker: 'mock-maker-subclass', member-accessor: 'member-accessor-reflection' }
-          - { mock-maker: 'mock-maker-inline', member-accessor: 'member-accessor-reflection' }
-
-    # All build steps
-    # SINGLE-MATRIX-JOB means that the step does not need to be executed on every job in the matrix
-    steps:
-
-    - name: 1. Check out code
-      uses: actions/checkout@v3 # https://github.com/actions/checkout
-      with:
-        fetch-depth: '0' # https://github.com/shipkit/shipkit-changelog#fetch-depth-on-ci
-
-    - name: 2. Set up Java ${{ matrix.java }}
-      uses: actions/setup-java@v3
-      with:
-        distribution: 'zulu'
-        java-version: ${{ matrix.java }}
-
-    - name: 3. Validate Gradle wrapper
-      if: matrix.java == 11 && matrix.entry.mock-maker == 'mock-maker-default' # SINGLE-MATRIX-JOB
-      uses: gradle/wrapper-validation-action@v1.0.6 # https://github.com/gradle/wrapper-validation-action
-
-    - name: 4. Build and check reproducibility of artifacts (single job only)
-      if: matrix.java == 11 && matrix.entry.mock-maker == 'mock-maker-default' # SINGLE-MATRIX-JOB
-      run: ./check_reproducibility.sh
-
-    - name: 5. Spotless check (single job only). Run './gradlew spotlessApply' locally if this job fails.
-      if: matrix.java == 11 && matrix.entry.mock-maker == 'mock-maker-default' # SINGLE-MATRIX-JOB
-      run: ./gradlew spotlessCheck
-
-    - name: 6. Build on Java ${{ matrix.java }} with ${{ matrix.entry.mock-maker }} and ${{ matrix.entry.member-accessor }}
-      run: ./gradlew build idea --scan
-      env:
-        MOCK_MAKER: ${{ matrix.entry.mock-maker }}
-        MEMBER_ACCESSOR: ${{ matrix.entry.member-accessor }}
-
-    - name: 7. Upload coverage report
-      run: |
-        ./gradlew coverageReport -s --scan && cp build/reports/jacoco/mockitoCoverage/mockitoCoverage.xml jacoco.xml
-        curl https://keybase.io/codecovsecurity/pgp_keys.asc | gpg --no-default-keyring --keyring trustedkeys.gpg --import # One-time step
-
-        curl -Os https://uploader.codecov.io/latest/linux/codecov
-        curl -Os https://uploader.codecov.io/latest/linux/codecov.SHA256SUM
-        curl -Os https://uploader.codecov.io/latest/linux/codecov.SHA256SUM.sig
-
-        gpgv codecov.SHA256SUM.sig codecov.SHA256SUM
-        shasum -a 256 -c codecov.SHA256SUM
-
-        chmod +x codecov
-        ./codecov
+  build-and-test:
+   uses: mockito/mockito/.github/workflows/build-and-test.yml@feature/build-once-test-all
```

Comment:
Why is this indirection necessary? I it looks like you extracted a file to call other files. Is it extracted so that there's no need to list all the *test* jobs in `needs` below?

Note: we can use [`matrix` with `uses`](https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-a-matrix-strategy-with-a-reusable-workflow), so theoretically you can say `needs: [test]` which refers to the whole test matrix.

Reply:
using matrix would mean test jobs will have to wait for ALL build jobs to finish. In reality it just needs to wait for a single build job in the matrix to finish. E.g testing on java 17  with default mock maker and member accessor, only needs to wait for build on java 11 with default mock maker and member accessor. It doesn't need to wait for other combination of mock maker and member accessor to complete.

or is it better to use matrix but it will be slower?

Reply:
Right now we have:
```mermaid
graph LR;
  build-default-default-->test11-default-default-->release
  build-inline-module-->test11-inline-module-->release
  build-inline-reflection-->test11-inline-reflection-->release
  build-subclass-module-->test11-subclass-module-->release
  build-subclass-reflection-->test11-subclass-reflection-->release
  build-default-default-->test17-default-default-->release
  build-inline-module-->test17-inline-module-->release
  build-inline-reflection-->test17-inline-reflection-->release
  build-subclass-module-->test17-subclass-module-->release
  build-subclass-reflection-->test17-subclass-reflection-->release
```
Based on a quick check there's only one variant of JAR build output. The mockmaker/member parameter is only used to set up test resources, so we should only need to build once, which means that we end up with this:
```mermaid
graph LR;
  build-->test11-default-default-->release
  build-->test11-inline-module-->release
  build-->test11-inline-reflection-->release
  build-->test11-subclass-module-->release
  build-->test11-subclass-reflection-->release
  build-->test17-default-default-->release
  build-->test17-inline-module-->release
  build-->test17-inline-reflection-->release
  build-->test17-subclass-module-->release
  build-->test17-subclass-reflection-->release
```
I think this means that the current structure can be kept and there's no need for workflow calls at all?

Reply:
Note that workflow calls have a really bad developer experience:
![image](https://github.com/mockito/mockito/assets/2906988/a743d3b1-2f10-496b-a7a6-9b4992a887ab)
compared to matrix:
![image](https://github.com/mockito/mockito/assets/2906988/05a1245a-245e-4644-8c60-c308748664cd)

Reply:
Can you please resolve this conversation now that you migrated to single file?
(Also the other files are still on the branch, did you want to keep those?)